examples/CircleOptimisation.lean

```diff
@@ -0,0 +1,104 @@
+/-
+Copyright (c) 2023 Siddharth Bhat. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Siddharth Bhat
+
+Riemannian Gradient Descent on S1.
+-/
+
+import SciLean
+open Lean
+open SciLean
+
+
+
+/-- choose a particular A -/
+-- the function (x, y) -> (x + y, x - y)
+def A (x : ℝ^{2}) : ℝ^{2} :=
+  ⊞ i, if i = 0 then x[0] + x[1] else x[0] - x[1]
+
+-- set_option trace.Meta.Tactic.fun_trans.rewrite true in
+function_properties A (x : ℝ^{2})
+argument x
+  IsLin, HasAdjoint, IsSmooth, HasAdjDiff,
```

Comment:
Where is `HasAdjoint` used by `SciLean`?

Reply:
All transformation rules for `†` are conditioned by `HasAdjoint`.

Similarly `∂` is conditioned by `IsSmooth`
and `∂†` is conditioned by `HasAdjDiff`

Because I'm working with generalization of Hilbert spaces, linearity of a map does not imply existence of adjoint. 
The canonical example is  constant function over natural numbers:
```
(λ (x : ℝ) (y : Nat) => x) : ℝ → (Nat → ℝ)
```
Morally adjoint is sum
```
sum : (Nat → ℝ) → ℝ 
```
but this is not well defined function as not every sequence is summable.

Why do I need this? I really want `(d/dx)† = -d/dx` where
```
d/dx : C∞(ℝ,ℝ) → C∞(ℝ,ℝ)
```
but `C∞(ℝ,ℝ)` is not Hilbert space